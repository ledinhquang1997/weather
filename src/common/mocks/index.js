export * from './MockData';
export * from './MockProvider';
