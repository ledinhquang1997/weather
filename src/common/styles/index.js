export * from './theme';
export * from './theme';
export * from './globalStyle';
