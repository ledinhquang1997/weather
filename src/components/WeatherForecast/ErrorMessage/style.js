import styled from 'styled-components';

export const Wrapper = styled.div`
  margin: 20px;
  background-color: rgba(0, 0, 0, 0.3);
  border-radius: 20px;
  padding: 5px 20px;
`;

export const Message = styled.p`
  color: ${({ theme }) => theme.colors.white};
`;
