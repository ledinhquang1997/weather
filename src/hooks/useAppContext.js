import { useContext } from 'react';

import { AppContext } from '../context';

export const useAppContext = () => {
  const { state, dispatch } = useContext(AppContext);

  return { state, dispatch };
};
