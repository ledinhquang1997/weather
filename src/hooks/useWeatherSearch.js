import { useContext, useState } from 'react';

import { search } from '../apis/weatherSearchApi';
import { validate } from '../common';
import { AppContext } from '../context';
import { SET_LOADING } from '../context/constants';

export const useWeatherSearch = () => {
  const [searchResult, setSearchResult] = useState([]);
  const [cityName, setCityName] = useState('');
  const [error, setError] = useState(null);
  const { dispatch } = useContext(AppContext);

  const handleSearch = async (searchText) => {
    try {
      const validateError = validate(searchText);

      if (validateError) {
        setError(validateError);

        return;
      }

      dispatch({ type: SET_LOADING, payload: true });
      const weather = await search(searchText);
      const result = weather?.data?.weather || [];
      const city = weather?.data?.city?.title;

      if (result.length <= 0 || !city) {
        setError('No city found :(');
      } else {
        setError(null);
        setCityName(city);
        setSearchResult(result);
      }

      dispatch({ type: SET_LOADING, payload: false });
    } catch (err) {
      setCityName('');
      setSearchResult([]);
      setError('No city found :(');
      dispatch({ type: SET_LOADING, payload: false });
    }
  };

  return { searchResult, handleSearch, error, cityName };
};
