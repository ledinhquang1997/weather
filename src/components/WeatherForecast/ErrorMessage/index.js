import { useAppContext } from '../../../hooks/useAppContext';
import { Message, Wrapper } from './style';

const ErrorMessage = () => {
  const {
    state: { error },
  } = useAppContext();

  return (
    error && (
      <Wrapper>
        <Message>{error}</Message>
      </Wrapper>
    )
  );
};

export default ErrorMessage;
