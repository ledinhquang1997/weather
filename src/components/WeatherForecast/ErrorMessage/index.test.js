import { render } from '@testing-library/react';

import ErrorMessage from '.';
import { MockProvider } from '../../../common';
import { useAppContext } from '../../../hooks/useAppContext';

jest.mock('../../../hooks/useAppContext');

describe('Error message componen', () => {
  test('shold render as expected', () => {
    useAppContext.mockImplementation(() => ({ state: { error: 'shold render as expected' } }));
    const { container } = render(
      <MockProvider>
        <ErrorMessage />
      </MockProvider>
    );

    expect(container.firstChild).toMatchSnapshot();
  });
});
