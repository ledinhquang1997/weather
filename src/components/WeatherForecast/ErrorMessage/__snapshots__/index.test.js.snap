// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error message componen shold render as expected 1`] = `
<div
  class="sc-bdnxRM cpWKGI"
>
  <p
    class="sc-gtsrHT gtXmqm"
  >
    shold render as expected
  </p>
</div>
`;
